### PR TITLE
[Mid-57] Feedback modal

### PR DESCRIPTION
**Pull request by Mykhailo-Pasternak**
# Feedback Form

This repository contains the code for a feedback form implementation with  Next.js. 

_**The main features of this project include:**_

- **Dynamic Feedback Form:** Implemented a dynamic feedback form that is displayed upon clicking the "Зворотній зв'язок" button.

- **Validation with Zod:** Utilized the "Zod" library for form validation, ensuring accurate and secure user input.

- **Data Submission:** Integrated functionality to send and store form data in a database upon submission.

- **Admin Panel:** Created an admin panel accessible at "/admin#/feedback" for convenient review of submitted feedback data.

Demo: https://www.loom.com/share/ca39593261d54d96bc5ae5d9705d8da2

LOCAL ENV DEMO PAGE: https://mayak-liart.vercel.app/

Admin panel:
https://mayak-liart.vercel.app/admin#/feedback

### **This task includes the following tasks:**
https://trello.com/c/AFnXPeGh/98-add-new-rules-for-lint

### CODE REVIEW

**Comment by watchlar2000** [comment]
Just a quick note after seeing the live demo, the errors shown up under the input field shouldn't move other input fields. You should consider using absolute positioning to display errors.

**Comment by Mykhailo-Pasternak** [comment]
@watchlar2000 то було демка, робочої логіки ( відпрацювання валідації та успішного відпавлення на БД ) Стилі лише вчора почав добавляти. 
Але так, ти вірно замітив, добре, було б, щоб вони не скакали... 
На жаль, той хто створював наш кастопний інпут, про це не піклувався.

@watchlar2000, that was a demo with the working logic (handling validation and successful database submission). I only started adding styles yesterday. But yes, you noticed correctly, it would be good if they didn't jump around. Unfortunately, the person who created our custom input didn't care about that.

File: .lintstagedrc.json [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
   "*.js": "eslint --ext .js,.mjs",
-  "*.{scss,md,json,prisma}": "prettier --check"
-}
+  "*.{scss,md,json}": "prettier --check"
```
Comment (alex-popov-tech) [inline .lintstagedrc.json]: 🟠  why not checking prisma?

File: src/app/_components/Feedback/Feedback.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
+import React, { useEffect, useState } from 'react';
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟠 
1. `React` is not needed i guess
2. component which consists of one file does not need a whole folder for it

File: src/app/_components/Feedback/Feedback.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import PropTypes from 'prop-types';
+import LoveIcon from '@icons/loveIcon.svg';
+import { useCreateFeedback } from '@/app/_hooks';
+import SendFeedback from '@/lib/validationSchemas/sendFeedbackSchema';
+import { Modal } from '../Modal';
+import { TextInputField } from '../InputFields';
+import { CheckBox } from '../CheckBox';
+import { TextArea } from '../TextArea';
+import { PillButton } from '../PillButton';
+import { cn } from '@/utils/cn';
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟠 imports should be organised as following:
1. packages
2. import aliases
3. relative imports

i think we even have a rule for this, don't we?

File: src/app/_components/Feedback/Feedback.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+import { buttonColorVariant, buttonVariant } from '../PillButton/style';
+
+const MAX_LENGTH = 320;
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟠 what is the point of having this const and using it only once? 🤔

File: src/app/_components/Feedback/Feedback.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  const [validationErrors, setValidationErrors] = useState({});
+
+  const { mutate: createFeedback } = useCreateFeedback();
+
+  const onSubmit = e => {
+    e.preventDefault();
+    const formData = {
+      name,
+      phone,
+      callMe: isChecked,
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟠 `isChecked` could be named better i think, in context of form 🤔

File: src/app/_components/Feedback/Feedback.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+import { CheckBox } from '../CheckBox';
+import { TextArea } from '../TextArea';
+import { PillButton } from '../PillButton';
+import { cn } from '@/utils/cn';
+import { buttonColorVariant, buttonVariant } from '../PillButton/style';
+
+const MAX_LENGTH = 320;
+
+export function Feedback({ isFeedbackOpen, onClose }) {
+  const [isFormOpen, setFormOpen] = useState(true);
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js:15]: 🟢 maybe better to do is otherwise - `isSubmitted` ?

File: src/app/_components/Feedback/Feedback.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+  const { mutate: createFeedback } = useCreateFeedback();
+
+  const onSubmit = e => {
+    e.preventDefault();
+    const formData = {
+      name,
+      phone,
+      callMe: isChecked,
+      email: email.length ? email : undefined,
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js:34]: 🟢 would be nice to have little comment about why is that needed

File: src/app/_components/Feedback/Feedback.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    try {
+      SendFeedback.parse(formData);
+      setValidationErrors({});
+      createFeedback(formData);
+
+      setFormOpen(false);
+    } catch (error) {
+      const errors = {};
+      error.errors.forEach(validationError => {
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟠 instead of mutating object in `forEach` consider using `map` and `reduce` to follow immutable approach

File: src/app/_components/Feedback/Feedback.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+              aria-label="Click to send feedback data"
+            >
+              Надіслати
+            </PillButton>
+          </form>
+        ) : (
+          <div className="grid justify-items-center">
+            <h3 className="pb-6 text-h3 font-bold">Дякую за повідомлення!</h3>
+            <p className="text-p2">Наші менеджери незабаром звʼяжуться з Вами</p>
+            <LoveIcon
```
Comment (alex-popov-tech) [inline src/app/_components/Feedback/Feedback.js]: 🟢 i would say its `Like` icon, not `Love` :)

File: src/app/_components/InputFields/TextInputField.js [hunk 10]
```diff
@@ -14,11 +14,15 @@ export function TextInputField({
   error = '',
   required = false,
   variant = variants.default,
+  absolute = true,
 }) {
   const id = `textinput_${name}`;
+  const absoluteError = absolute ? 'absolute top-[45px] duration-150 ease-out transform' : '';
```
Comment (alex-popov-tech) [inline src/app/_components/InputFields/TextInputField.js:20]: 🟠 are these `absolute` pixels are working alright in mobile view too?

File: src/app/_components/Feedback/Feedback.js (hunk 4 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: I don't like having magic numbers in my code.

File: src/app/_components/Feedback/Feedback.js (hunk 9 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: Це назва із дизайну) Можу обізвати ту іконку як завгодно))

File: src/app/_components/InputFields/TextInputField.js (hunk 10 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/InputFields/TextInputField.js:20]: я можу гянути, але  глянь у мій тікет, що від мене вимагається зробити)) 
я мав робити для декстопу

File: src/app/_components/Feedback/Feedback.js (hunk 4 above)
Comment (alex-popov-tech) [reply src/app/_components/Feedback/Feedback.js]: do you think `maxLegth={MAX_LENGTH}` helps with that? :)

Comment (watchlar2000) [reply src/app/_components/Feedback/Feedback.js]: Rather than just being some integer 320, assigning it to MAX_LENGTH variable describes what it actually means and will help you understand code better when you come back later.

File: src/app/_components/Feedback/Feedback.js (hunk 7 above)
Comment (watchlar2000) [reply src/app/_components/Feedback/Feedback.js:34]: Just an alternative option:
```js
email: email?.length // either email.length or undefined
```

File: src/app/_components/Feedback/Feedback.js (hunk 3 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: **If this condition is really important, then create a ticket.**

_The build of the application with these imports is successful and does not cause any problems._

File: .lintstagedrc.json (hunk 1 above)
Comment (Mykhailo-Pasternak) [reply .lintstagedrc.json]: Хаскі не вміє працювати із "prisma", та видає дає помилку, ПОЗБАВЛЯЄ можливості відправити свій код до github. 
I без цих змін, я \ чи будь хто, хто працює на операційній системі віндовс, змушений видаляти "prisma" із правил, відправляти код і тоді вертає назад, це правило.  Такий костиль, нам без сенсу підтримувати.

"Husky doesn't work with Prisma and throws an error, preventing the ability to push the code to GitHub. Without these changes, I, or anyone working on the Windows operating system, have to exclude 'Prisma' from the rules, push the code, and then revert this rule. This workaround makes no sense for us to maintain."

File: src/app/_components/Feedback/Feedback.js (hunk 6 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js:15]: IF **isFormOpen === true**   _I show Feedback form_

IF **isFormOpen === false** _I hide form and show "Thank You Modal"_

File: .lintstagedrc.json (hunk 1 above)
Comment (alex-popov-tech) [reply .lintstagedrc.json]: true about workaround

altho deleting a prisma file from linting is not an option too, as that leads to degradating codebase ( in this case, db schema )

i suggest we ( or anybody else ) try to pair on that, and fix the actual issue with editor on windows, rather than changing part of codebase in favor of editor which is not setup correctly 

if nothing can be done, i would expect an explanation of that somewhere in the docs 🤷🏻‍♂️

File: src/app/_components/Feedback/Feedback.js (hunk 3 above)
Comment (alex-popov-tech) [reply src/app/_components/Feedback/Feedback.js]: true, im sure i saw linting errors about that in my code and was fixing them, not sure why now i don't see them 🤔 
anyway, comment resolved

File: src/app/_components/Feedback/Feedback.js (hunk 8 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: **Is the mutable approach bad in this case?**

_1 forEach - It simply iterates over the elements. **( My case )**
2  map: It creates a new array by applying a function to each element of the original array. 
3 reduce:  when you need to accumulate a result._

File: src/app/_components/Feedback/Feedback.js (hunk 4 above)
Comment (alex-popov-tech) [reply src/app/_components/Feedback/Feedback.js]: could not agree to that completely :)

seeing the code

```
...
maxLength={320}
...
```

means 'there is an max length argument, and im passing 320' - so we kinda can make assumption that 320 _is_ max length

in this code

```
...
maxLength={MAX_LENGTH}
...
```

we can take 'there is max length argument, and max length constant passed in' - where second part is kinda makes no sense, as it is not adding anything about 'what it is actually are', there are no additional sense in there

what could be done ( if still wanted ) is something like
```
...
maxLength={MAX_MESSAGE_LENGTH}
...
```
as this adds some bits of information why there is 320 - because it is a 'message' and business rules are that it has to be <=320

File: src/app/_components/Feedback/Feedback.js (hunk 8 above)
Comment (alex-popov-tech) [reply src/app/_components/Feedback/Feedback.js]: it is a general rule/practise that any kind of mutation is 'bad' - as it is complicated, and leaves less traces to debug, comparing to immutable approach

you could also say that from philosophical point of view that 'forEach' is _consuming_ data, rather than used as an any iterator to do code you want inside on each item

map/reduce on the other hand are data modificators - they are suppose to change shape of data, so these are preferred to use when you want to transform data from one shape to another

File: .lintstagedrc.json (hunk 1 above)
Comment (Mykhailo-Pasternak) [reply .lintstagedrc.json]: You're right. I will return the rule with a prism.

File: src/app/_components/Feedback/Feedback.js (hunk 4 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: You gave a comprehensive answer. I agree.

File: src/app/_components/Feedback/Feedback.js (hunk 7 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js:34]: **_added an explanatory comment to the code:_**  
```
 // I am forced to do exactly that, since I have to pass a valid email or "undefined" to the backend.
 // Input should have base value ""
```

File: src/app/_components/Feedback/Feedback.js (hunk 8 above)
Comment (Mykhailo-Pasternak) [reply src/app/_components/Feedback/Feedback.js]: Аргументована відповідь. 
Добавив відповідні зміни.

File: package.json [hunk 11]
```diff
@@ -56,6 +56,9 @@
     "eslint-config-airbnb-base": "^15.0.0",
     "eslint-config-next": "14.1.0",
     "eslint-config-prettier": "^9.1.0",
+    "eslint-import-resolver-alias": "^1.1.2",
```
Comment (alex-popov-tech) [inline package.json]: I think we don't need resolves, do we?

**Review by alex-popov-tech — APPROVED** [review]

File: package.json (hunk 11 above)
Comment (Mykhailo-Pasternak) [reply package.json]: **_deleted redundant dependencies_**

File: src/app/_components/Header.js [hunk 12]
```diff
@@ -64,6 +71,7 @@ export function Header() {
             className={cn(basicLink, transition, iconColors, 'hover:color-primary-500 gap-6 hover:text-primary-500')}
           />
           <OutlinedButton
```
Comment (Yivgha) [inline src/app/_components/Header.js:73]: 🟢  PillButton must be here, not OutlinedButton.

File: src/app/_components/InputFields/TextInputField.js [hunk 13]
```diff
@@ -14,11 +14,15 @@ export function TextInputField({
...
+  const absoluteLabel = absolute ? 'absolute bottom-[47px]' : '';
+
   return (
-    <div className={cn(variant.mainContainer.base)}>
-      {error && <p className={cn(variant.errorParagraph.base)}>{error}</p>}
+    <div className={cn(`${variant.mainContainer.base} relative`)}>
```
Comment (Yivgha) [inline src/app/_components/InputFields/TextInputField.js:24]: 🟠 if you're using cn utility, consider to refactor styles like this: className={cn("variant.mainContainer.base relative")}

Comment (watchlar2000) [reply src/app/_components/InputFields/TextInputField.js:24]: guys, why do you need "cn" here at all?

Comment (Mykhailo-Pasternak) [reply src/app/_components/InputFields/TextInputField.js:24]: @Yivgha, I apologize. But the variables:
`variant.mainContainer.base`
in your use case will become just a string.

Because I use (Template strings) as  `${variant.mainContainer.base} relative`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Comment (Mykhailo-Pasternak) [reply src/app/_components/InputFields/TextInputField.js:24]: @watchlar2000 to be honest, I don't see the point either))
in the above code, I just added the "relative" style

**Review by Yivgha — APPROVED** [review]
